frontend/src/utils/address.ts

```diff
@@ -154,11 +153,65 @@ export const lookupOrReturnAddresses = async (addresses: string[], provider: Pro
     if (!success || returnData === '0x') return names[i]; // This is an address in these cases.
     return <string>resolverInterface.decodeFunctionResult('addr', returnData)[0];
   });
+  return { names, forwardAddrs };
+};
+
+const lookupCNSNameBatch = async (
+  addresses: string[],
+  registryAddr: string,
+  provider: Provider | StaticJsonRpcProvider
+) => {
+  const multicall = new Contract(MULTICALL_ADDRESS, MULTICALL_ABI, provider);
+  const ensRegistryInterface = new Interface(['function reverseNameOf(address) view returns (string)']);
+  const reverseResolverCalls = addresses.map((addr) => ({
+    target: registryAddr,
+    allowFailure: true,
+    callData: ensRegistryInterface.encodeFunctionData('reverseNameOf', [addr]),
+  }));
+
+  type Response = { success: boolean; returnData: string };
+  const reverseResolverResults: Response[] = await multicall.callStatic.aggregate3(reverseResolverCalls);
+  const results = reverseResolverResults.map((resp) => {
+    if (resp?.returnData !== AddressZero) {
+      const addr = defaultAbiCoder.decode(['string'], resp.returnData) as string[];
+      return addr[0];
+    }
+    return '';
+  });
+  return results;
+};
+
+const lookupCNSNameBatchMainnet = async (addresses: string[]) => {
+  return await lookupCNSNameBatch(addresses, '0xCd451149ffa9d059030528917842bcE14327DfD6', MAINNET_PROVIDER);
+};
+
+const lookupCNSNameBatchPolygon = async (addresses: string[]) => {
+  return await lookupCNSNameBatch(addresses, '0xa9a6A3626993D487d2Dbda3173cf58cA1a9D9e9f', POLYGON_PROVIDER);
+};
+
+const lookupCnsNameBatch = async (addresses: string[]) => {
+  try {
+    const [resultL1, resultL2] = await Promise.all([
+      lookupCNSNameBatchMainnet(addresses),
+      lookupCNSNameBatchPolygon(addresses),
+    ]);
```

Comment:
Is this correct: a name can be registered on mainnet OR polygon only, so we check both, and those are the only two chains you can register on which is why we don't check any others?

Reply:
Yes, outside of testnets. I based a lot of this off of their resolution library. It was unfortunate they didn't support batching out of the box.

https://docs.unstoppabledomains.com/smart-contracts/contract-reference/uns-smart-contracts/#unsregistry